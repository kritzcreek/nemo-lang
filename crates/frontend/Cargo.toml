[package]
name = "frontend"
version = "0.1.0"
edition = "2021"
license = "MIT"

[dependencies]
ariadne = "0.4.1"
yansi = "1"
logos = "0.14.0"
num-derive = "0.4.2"
num-traits = "0.2.18"
rowan = "0.15.15"
text-size = "1.1.1"
line-index = "0.1.1"
lexpr = "0.2.7"
derive_ir = { path = "../derive_ir" }

[dev-dependencies]
insta = { version = "1.34.0", features = ["glob"] }
