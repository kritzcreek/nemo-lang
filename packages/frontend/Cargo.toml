[package]
name = "frontend"
version = "0.1.0"
edition = "2021"
license = "MIT"

[dependencies]
ariadne = "0.4.0"
yansi = "0.5.1"
logos = "0.14.0"
num-derive = "0.4.2"
num-traits = "0.2.18"
rowan = "0.15.15"
lazy_static = "1.4.0"
backend = { path = "../backend" }
text-size = "1.1.1"
line-index = "0.1.1"

[dev-dependencies]
insta = { version = "1.34.0", features = ["glob"] }
