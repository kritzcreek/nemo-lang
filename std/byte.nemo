module byte
exports (eq, concat)

fn eq(a: bytes, b: bytes) -> bool {
  let len = bytes_len(a);
  if len != bytes_len(b) {
    return false;
  } else {};
  let i = 0;
  while i < len {
    if bytes_get(a, i) != bytes_get(b, i) {
      return false;
    } else {};
    set i = i + 1;
  };
  true
}

fn concat(b1: bytes, b2: bytes) -> bytes {
  let len1 = bytes_len(b1);
  let len2 = bytes_len(b2);
  let len = len1 + len2;
  let b = bytes_new(0, len);
  bytes_copy(b, 0, b1, 0, len1);
  bytes_copy(b, len1, b2, 0, len2);
  b
}
