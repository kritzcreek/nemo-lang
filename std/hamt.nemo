module hamt
exports (Hamt, new, get, add, remove, values, print)

use option
use vec
use string

fn print_hash(hash : I32) -> string::Str {
  let chars : vec::Vec[I32] = vec::new(0);
  let i = 0;
  while i < 32 {
    if i != 0 && i32_rem_u(i, 5) == 0 {
      vec::push(chars, 95)
    } else {};
    if i32_and(hash, 1) == 0 { vec::push(chars, 48) } else { vec::push(chars, 49) };
    set hash = i32_shr_u(hash, 1);
    set i = i + 1;
  };

  let char_count = chars.len;
  let bytes = bytes_new(0, char_count);
  let j = 0;
  while j < char_count {
    bytes_set(bytes, char_count - j - 1, chars.buf[j]);
    set j = j + 1;
  };
  string::unsafe_new(bytes)
}

fn print_bitmap(bitmap : I32) -> string::Str {
  let chars : vec::Vec[I32] = vec::new(0);
  let i = 0;
  while i < 32 {
    if i32_and(bitmap, 1) == 0 { vec::push(chars, 48) } else { vec::push(chars, 49) };
    set bitmap = i32_shr_u(bitmap, 1);
    set i = i + 1;
  };

  let char_count = chars.len;
  let bytes = bytes_new(0, char_count);
  let j = 0;
  while j < char_count {
    bytes_set(bytes, char_count - j - 1, chars.buf[j]);
    set j = j + 1;
  };
  string::unsafe_new(bytes)
}

fn print[a](hamt : Hamt[a]) -> string::Str {
  print_node#[a](hamt.root)
}

fn print_node[a](node : Node[a]) -> string::Str {
  match node {
    Node::Leaf l => {
      string::concat(string::concat(string::unsafe_new("(#leaf "), print_hash(l.hash)), string::unsafe_new(")"));
    },
    Node::Mapped m => {
      let nodes = string::unsafe_new("");
      let i = 0;
      while i < array_len#[Node[a]](m.nodes) {
        set nodes = string::concat(string::concat(nodes, string::unsafe_new(" ")), print_node(m.nodes[i]));
        set i = i + 1;
      };
      string::concat(
        string::concat(
          string::concat(
            string::unsafe_new("(#mapd "),
            print_bitmap(m.bitmap)),
          nodes),
        string::unsafe_new(")"));
    },
  };
}

struct Hamt[a] {
  root : Node::Mapped[a],
  size : I32,
}

variant Node[a] {
  struct Mapped { bitmap : I32, nodes : Array[Node[a]] },
  struct Leaf { hash : I32, value : a },
}

fn new[a]() -> Hamt[a] {
  Hamt { root = Node::Mapped { bitmap = 0, nodes = [] }, size = 0 }
}

fn get[a](hamt : Hamt[a], hash : I32) -> option::Option[a] {
  match get_with_anchor(hamt.root, 0, hash).payload {
    GetResultPayload::Success s => {
      option::Option::Some { val = s.node.value }
    },
    GetResultPayload::Missing _ => { option::Option::None {} },
    GetResultPayload::Conflict _ => { option::Option::None {} },
  };
}

fn add[a](hamt : Hamt[a], hash : I32, value : a) -> option::Option[a] {
  let get_result = get_with_anchor(hamt.root, 0, hash);
  match get_result.payload {
    GetResultPayload::Missing _ => {
      set hamt.size = hamt.size + 1;
      let n = get_result.anchor;
      let pos = bitpos(hash, get_result.shift);
      set n.bitmap = i32_or(n.bitmap, pos);
      let ix = index(n.bitmap, pos);
      set n.nodes = insert_node(n.nodes, ix, Node::Leaf { hash = hash, value = value });
      option::Option::None {};
    },
    GetResultPayload::Conflict c => {
      set hamt.size = hamt.size + 1;
      let n = get_result.anchor;
      let pos = bitpos(hash, get_result.shift);
      let ix = index(n.bitmap, pos);
      let new_node = merge_leafs(get_result.shift + 5, c.node, hash, value);
      set n.nodes[ix] = new_node;
      option::Option::None {};
    },
    GetResultPayload::Success s => {
      let prev = s.node.value;
      set s.node.value = value;
      option::Option::Some { val = prev };
    },
  };
}

fn remove[a](hamt : Hamt[a], hash : I32) -> option::Option[a] {
  match remove_rec(hamt.root, 0, hash) {
    RemoveResult::Success s => {
      set hamt.size = hamt.size - 1;
      option::Option::Some { val = s.node.value }
    },
    RemoveResult::Gathered g => {
      set hamt.size = hamt.size - 1;
      option::Option::Some { val = g.removed.value }
    },
    RemoveResult::NotFound n => { option::Option::None {} },
  };
}

fn find_any[a](node : Node[a]) -> option::Option[a] {
  match node {
    Node::Leaf l => { option::Option::Some { val = l.value } },
    Node::Mapped m => {
      let i = 0;
      let len = array_len#[Node[a]](m.nodes);
      while i < len {
        match find_any(m.nodes[i]) {
          option::Option::Some s => { return s },
          option::Option::None _ => { },
        };
        set i = i + 1;
      };
      option::Option::None {};
    },
  };
}

fn values[a](hamt : Hamt[a]) -> option::Option[vec::Vec[a]] {
  let zero = find_any#[a](hamt.root);
  let result = match zero {
    option::Option::Some s => { vec::new(s.val) },
    option::Option::None _ => { return option::Option::None {} },
  };
  values_node(result, hamt.root);
  option::Option::Some { val = result };
}

fn values_node[a](vec : vec::Vec[a], node : Node[a]) {
  match node {
    Node::Leaf l => { vec::push(vec, l.value) },
    Node::Mapped m => {
      let i = 0;
      let len = array_len#[Node[a]](m.nodes);
      while i < len {
        values_node(vec, m.nodes[i]);
        set i = i + 1;
      };
    },
  };
}

fn mask(hash : I32, shift : I32) -> I32 {
  i32_and(i32_shr_u(hash, shift), 31)
}

fn bitpos(hash : I32, shift : I32) -> I32 {
   i32_shl(1, mask(hash, shift))
}

fn index(bitmap : I32, pos : I32) -> I32 {
  i32_popcnt(i32_and(bitmap, (pos - 1)))
}

fn merge_leafs[a](shift : I32, left : Node::Leaf[a], hash : I32, value : a) -> Node[a] {
  let next_pos1 = bitpos(left.hash, shift);
  let next_pos2 = bitpos(hash, shift);
  if next_pos1 != next_pos2 {
    let bitmap = i32_or(next_pos1, next_pos2);
    let right = Node::Leaf { hash = hash, value = value };
    let nodes : Array[Node[a]] = if i32_lt_u(next_pos1, next_pos2) {
      [left, right]
    } else {
      [right, left]
    };
    Node::Mapped { bitmap = bitmap, nodes = nodes };
  } else {
    let new_node = merge_leafs(shift + 5, left, hash, value);
    Node::Mapped { bitmap = next_pos1, nodes = [new_node] };
  };
}

variant RemoveResult[a] {
  struct Success {
    node : Node::Leaf[a],
  },
  struct NotFound {},
  struct Gathered {
    new_node : Node::Leaf[a],
    removed : Node::Leaf[a],
  },
}

fn remove_rec[a](node : Node::Mapped[a], shift : I32, hash : I32) -> RemoveResult[a] {
  let pos = bitpos(hash, shift);
  if i32_and(pos, node.bitmap) == 0 {
    return RemoveResult::NotFound {};
  } else {};
  let ix = index(node.bitmap, pos);
  let next_node = node.nodes[ix];
  match next_node {
    Node::Mapped n => {
      let res = remove_rec(n, shift + 5, hash);
      match res {
        RemoveResult::Gathered gathered => {
          if shift != 0 && i32_popcnt(n.bitmap) == 1 {
            res;
          } else {
            set n.nodes[ix] = gathered.new_node;
            RemoveResult::Success { node = gathered.removed };
          };
        },
        _ => { res },
      };
    },
    Node::Leaf l => {
      if hash != l.hash {
        RemoveResult::NotFound {};
      } else {
        let rows = i32_popcnt(node.bitmap);
        // We never gather the root node
        if rows == 1 && shift == 0 {
          set node.bitmap = 0;
          set node.nodes = [];
          return RemoveResult::Success { node = l };
        } else {};
        if rows == 2 {
          let other = if ix == 0 { node.nodes[1] } else { node.nodes[0] };
          match other {
            Node::Leaf other_l => {
              if shift != 0 {
                return RemoveResult::Gathered { new_node = other_l, removed = l };
              } else {};
            },
            _ => {},
          };
          set node.bitmap = i32_and(node.bitmap, i32_xor(pos, 0 - 1));
          set node.nodes = [other];
          RemoveResult::Success { node = l };
        } else {
          let new_nodes = remove_node(node.nodes, ix);
          set node.bitmap = i32_and(node.bitmap, i32_xor(pos, 0 - 1));
          set node.nodes = new_nodes;
          RemoveResult::Success { node = l };
        };
      };
    },
  };
}

fn remove_node[a](as : Array[a], ix : I32) -> Array[a] {
  let old_len = array_len#[a](as);
  if old_len == 1 {
    [];
  } else {
    let res = array_new#[a](as[0], old_len - 1);
    array_copy#[a](res, 0, as, 0, ix);
    array_copy#[a](res, ix, as, ix + 1, old_len - ix - 1);
    res;
  };
}

fn insert_node[a](as : Array[a], ix : I32, elem : a) -> Array[a] {
  let old_len = array_len#[a](as);
  if old_len == 0 {
    [elem];
  } else {
    let res = array_new#[a](elem, old_len + 1);
    array_copy#[a](res, 0, as, 0, ix);
    array_copy#[a](res, ix + 1, as, ix, old_len - ix);
    res;
  };
}

struct GetResult[a] {
  shift : I32,
  anchor : Node::Mapped[a],
  payload : GetResultPayload[a],
}

variant GetResultPayload[a] {
  struct Success { node : Node::Leaf[a] },
  struct Conflict { node : Node::Leaf[a] },
  struct Missing {},
}

fn get_with_anchor[a](
  anchor : Node::Mapped[a],
  shift : I32,
  hash : I32,
) -> GetResult[a] {
  let pos = bitpos(hash, shift);
  if i32_and(anchor.bitmap, pos) == 0 {
    GetResult {
      shift = shift,
      anchor = anchor,
      payload = GetResultPayload::Missing {},
    };
  } else {
    let ix = index(anchor.bitmap, pos);
    let n = anchor.nodes[ix];
    match n {
      Node::Mapped anchor => {
        get_with_anchor(anchor, shift + 5, hash);
      },
      Node::Leaf leaf => {
        if leaf.hash == hash {
          GetResult {
            shift = shift,
            anchor = anchor,
            payload = GetResultPayload::Success { node = leaf },
          };
        } else {
          GetResult {
            shift = shift,
            anchor = anchor,
            payload = GetResultPayload::Conflict { node = leaf },
          };
        };
      },
    };
  };
}
