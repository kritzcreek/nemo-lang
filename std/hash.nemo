module hash
exports (FnvHasher, newHasher, write, write_i32, hash_i32, hash_bytes)

global fnv_prime = 16777619
// 2166136261 as a signed integer
global fnv_offset_basis = 0 - 2128831035

// http://www.isthe.com/chongo/tech/comp/fnv/#FNV-1a
fn fnv_1a(hash : I32, input : Bytes) -> I32 {
  let len = bytes_len(input);
  let i = 0;
  while i < len {
    set hash = i32_xor(hash, bytes_get(input, i)) * fnv_prime;
    set i = i + 1;
  };
  hash
}

fn fnv_1a_i32(hash : I32, input : I32) -> I32 {
  set hash = i32_xor(hash, i32_and(input, 0xFF)) * fnv_prime;
  set hash = i32_xor(hash, i32_and(i32_shr_u(input, 8), 0xFF)) * fnv_prime;
  set hash = i32_xor(hash, i32_and(i32_shr_u(input, 16), 0xFF)) * fnv_prime;
  i32_xor(hash, i32_shr_u(input, 24)) * fnv_prime
}

fn hash_i32(i : I32) -> I32 {
  fnv_1a_i32(fnv_offset_basis, i)
}

fn hash_bytes(i : Bytes) -> I32 {
  fnv_1a(fnv_offset_basis, i)
}

struct FnvHasher { hash : I32 }

fn newHasher() -> FnvHasher {
  FnvHasher { hash = fnv_offset_basis }
}

fn write(hasher : FnvHasher, bytes : Bytes) {
  set hasher.hash = fnv_1a(hasher.hash, bytes)
}

fn write_i32(hasher : FnvHasher, int : I32) {
  set hasher.hash = fnv_1a_i32(hasher.hash, int)
}
