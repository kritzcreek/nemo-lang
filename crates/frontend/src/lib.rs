pub mod builtins;
pub mod highlight;
pub mod ir;
pub mod parser;
pub mod syntax;
pub mod types;

use crate::ir::NameMap;
use line_index::{LineCol, LineIndex};
use parser::{parse_prog, ParseError};
use std::fmt::{self, Write};
use text_size::TextRange;
use types::{CheckResult, TyError};

#[derive(Debug)]
pub enum CheckError {
    ParseError(ParseError),
    TypeError(TyError),
}

impl CheckError {
    pub fn display<'src, 'err>(
        &'err self,
        source: &'src str,
        name_map: &'src NameMap,
        colors: bool,
    ) -> DisplayCheckError<'src, 'err> {
        DisplayCheckError {
            source,
            error: self,
            name_map,
            colors,
        }
    }

    pub fn line_col(&self, line_index: &LineIndex) -> (LineCol, LineCol) {
        match self {
            CheckError::ParseError(err) => err.line_col(line_index),
            CheckError::TypeError(err) => err.line_col(line_index),
        }
    }

    pub fn message(&self, name_map: &NameMap) -> String {
        match self {
            CheckError::ParseError(err) => err.it.clone(),
            CheckError::TypeError(err) => err.message(name_map),
        }
    }

    pub fn at(&self) -> TextRange {
        match self {
            CheckError::ParseError(err) => err.at,
            CheckError::TypeError(err) => err.at,
        }
    }
}

pub struct DisplayCheckError<'a, 'b> {
    error: &'b CheckError,
    source: &'a str,
    name_map: &'a NameMap,
    colors: bool,
}

impl fmt::Display for DisplayCheckError<'_, '_> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self.error {
            CheckError::ParseError(err) => {
                write!(f, "{}", err.display(self.source, self.colors))
            }
            CheckError::TypeError(err) => {
                write!(
                    f,
                    "{}",
                    err.display(self.source, self.name_map, self.colors)
                )
            }
        }
    }
}

pub fn render_errors(errs: &[CheckError], source: &str, name_map: &NameMap) -> String {
    let mut output = String::new();
    for err in errs {
        writeln!(output, "{}", err.display(source, name_map, true)).unwrap();
    }
    output
}

/// Runs the full frontend on `source` and returns the generated IR and other structures.
/// If there are any errors, the generated IR should _not_ be used. It's returned here for
/// debugging purposes.
pub fn run_frontend(source: &str) -> CheckResult<CheckError> {
    let parse = parse_prog(source);
    let mut errors = vec![];
    let check_result = types::check_prog(parse.root());

    for error in parse.errors {
        errors.push(CheckError::ParseError(error));
    }
    for error in check_result.errors {
        errors.push(CheckError::TypeError(error));
    }
    if errors.is_empty() && check_result.ir.is_none() {
        panic!("No IR generated, despite no errors")
    };

    CheckResult {
        errors,
        names: check_result.names,
        occurrences: check_result.occurrences,
        ir: check_result.ir,
        parse: check_result.parse,
    }
}

/// Checks the given program, and prints any parse or type errors.
/// If there are any errors returns a summary message in Err otherwise Ok.
pub fn check_program(source: &str) -> Result<(), String> {
    let check_result = run_frontend(source);
    if !check_result.errors.is_empty() {
        for err in &check_result.errors {
            eprintln!(
                "{}",
                err.display(source, &check_result.names.name_map, true)
            );
        }
        return Err(format!(
            "Check failed with {} errors",
            check_result.errors.len()
        ));
    }
    Ok(())
}
