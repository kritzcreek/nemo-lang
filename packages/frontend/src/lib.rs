pub mod builtins;
pub mod highlight;
pub mod lexer;
pub mod parser;
pub mod syntax;
pub mod types;

use backend::{
    codegen::codegen,
    ir::{NameMap, NameSupply},
};
use line_index::{LineCol, LineIndex};
use parser::{parse_prog, ParseError};
use rowan::TextRange;
use std::fmt::{self, Write};
use syntax::{ast::AstNode, nodes::Root};
use types::{errors::TyError, CheckResult};

#[derive(Debug)]
pub enum CheckError {
    ParseError(ParseError),
    TypeError(TyError),
}

impl CheckError {
    pub fn display<'src, 'err>(
        &'err self,
        source: &'src str,
        name_map: &'src NameMap,
        colors: bool,
    ) -> DisplayCheckError<'src, 'err> {
        DisplayCheckError {
            source,
            error: self,
            name_map,
            colors,
        }
    }

    pub fn line_col(&self, line_index: &LineIndex) -> (LineCol, LineCol) {
        match self {
            CheckError::ParseError(err) => err.line_col(line_index),
            CheckError::TypeError(err) => err.line_col(line_index),
        }
    }

    pub fn message(&self, name_map: &NameMap) -> String {
        match self {
            CheckError::ParseError(err) => err.it.clone(),
            CheckError::TypeError(err) => err.message(name_map),
        }
    }

    pub fn at(&self) -> TextRange {
        match self {
            CheckError::ParseError(err) => err.at,
            CheckError::TypeError(err) => err.at,
        }
    }
}

pub struct DisplayCheckError<'a, 'b> {
    error: &'b CheckError,
    source: &'a str,
    name_map: &'a NameMap,
    colors: bool,
}

impl fmt::Display for DisplayCheckError<'_, '_> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self.error {
            CheckError::ParseError(err) => {
                write!(f, "{}", err.display(self.source, self.colors))
            }
            CheckError::TypeError(err) => {
                write!(
                    f,
                    "{}",
                    err.display(self.source, self.name_map, self.colors)
                )
            }
        }
    }
}

pub fn render_errors(errs: &[CheckError], source: &str, name_map: &NameMap) -> String {
    let mut output = String::new();
    for err in errs {
        writeln!(output, "{}", err.display(source, name_map, true)).unwrap();
    }
    output
}

pub fn run_frontend(source: &str) -> CheckResult<CheckError> {
    let parse = parse_prog(source);
    let mut errors = vec![];
    let check_result = match Root::cast(parse.syntax()) {
        None => panic!("Parse didn't yield a Root node"),
        Some(root) => types::check_prog(root),
    };

    for error in parse.errors {
        errors.push(CheckError::ParseError(error));
    }
    for error in check_result.errors {
        errors.push(CheckError::TypeError(error));
    }
    if errors.is_empty() && check_result.ir.is_none() {
        panic!("No IR generated, despite no errors")
    };

    CheckResult {
        errors,
        names: check_result.names,
        typed_nodes: check_result.typed_nodes,
        occurrences: check_result.occurrences,
        ir: check_result.ir,
        parse: check_result.parse,
    }
}

pub fn check_program(source: &str) -> (NameSupply, Vec<CheckError>) {
    let check_result = run_frontend(source);
    (check_result.names, check_result.errors)
}

pub fn compile_program(source: &str) -> (NameSupply, Result<Vec<u8>, Vec<CheckError>>) {
    let check_result = run_frontend(source);
    match check_result.ir {
        Some(ir) if check_result.errors.is_empty() => {
            let (wasm, names) = codegen(ir, check_result.names);
            (names, Ok(wasm))
        }
        _ => (check_result.names, Err(check_result.errors)),
    }
}
