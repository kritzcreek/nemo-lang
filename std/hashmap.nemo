module hashmap
exports (HashMap, new, add, get)

use io
use hamt
use option

variant Bucket[k, v] {
  struct Single { key : k, value : v },
  struct Conflict { key : k, value : v, next : Bucket[k, v] },
}

fn bucket_key[k, v](b : Bucket[k, v]) -> k {
  match b {
    Bucket::Single s => { s.key },
    Bucket::Conflict c => { c.key },
  };
}

fn bucket_value[k, v](b : Bucket[k, v]) -> v {
  match b {
    Bucket::Single s => { s.value },
    Bucket::Conflict c => { c.value },
  };
}

fn bucket_set_value[k, v](b : Bucket[k, v], value : v) {
  match b {
    Bucket::Single s => { set s.value = value },
    Bucket::Conflict c => { set c.value = value },
  };
}

fn bucket_add[k, v](b : Bucket[k, v], key : k, value : v) -> Bucket[k, v] {
  Bucket::Conflict { key = key, value = value, next = b }
}

fn bucket_find[k, v](
  b : Bucket[k, v],
  key : k,
  eq : fn (k, k) -> Bool
) -> option::Option[Bucket[k, v]] {
  match b {
    Bucket::Single single => {
      if eq(single.key, key) {
        option::Option::Some { val = single };
      } else {
        option::Option::None {};
      }
    },
    Bucket::Conflict conflict => {
      if eq(conflict.key, key) {
        option::Option::Some { val = conflict };
      } else {
        bucket_find(conflict.next, key, eq)
      }
    },
  }
}

fn bucket_get[k, v](
  b : Bucket[k, v],
  key : k,
  eq : fn (k, k) -> Bool
) -> option::Option[v] {
  option::map(bucket_find(b, key, eq), \(b) { bucket_value(b) })
}

struct HashMap[k, v] {
  hash : fn (k) -> I32,
  eq : fn (k, k) -> Bool,
  hamt : hamt::Hamt[Bucket[k, v]],
  size : I32,
}

fn new[k, v](hash : fn (k) -> I32, eq : fn (k, k) -> Bool) -> HashMap[k, v] {
  HashMap { hash = hash, eq = eq, hamt = hamt::new(), size = 0 }
}

fn add[k, v](map : HashMap[k, v], key : k, value : v) -> option::Option[v] {
  let hash = map.hash(key);
  match hamt::get(map.hamt, hash) {
    option::Option::Some s => {
      match bucket_find(s.val, key, map.eq) {
        option::Option::Some conflict => {
          let old_value = bucket_value(conflict.val);
          bucket_set_value(conflict.val, value);
          option::Option::Some { val = old_value }
        },
        option::Option::None _ => {
          hamt::add(map.hamt, hash, bucket_add(s.val, key, value));
          set map.size = map.size + 1;
          option::Option::None {};
        }
      };
    },
    option::Option::None _ => {
      hamt::add(map.hamt, hash, Bucket::Single { key = key, value = value });
      set map.size = map.size + 1;
      option::Option::None {}
    },
  };
}

fn get[k, v](map : HashMap[k, v], key : k) -> option::Option[v] {
  let hash = map.hash(key);
  match hamt::get(map.hamt, hash) {
    option::Option::Some bucket => {
      bucket_get(bucket.val, key, map.eq)
    },
    option::Option::None _ => {
      option::Option::None {};
    },
  };
}
