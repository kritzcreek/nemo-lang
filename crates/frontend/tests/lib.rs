use frontend::run_frontend;
use frontend::{parser::parse_prog, CheckError};
use insta::{assert_snapshot, glob};
use std::fmt::Write;
use std::fs;
use std::path::Path;
use std::str;

/// Copied from the Rust compiler: https://github.com/rust-lang/rust/pull/62948/
/// Replaces `\r\n` with `\n` in-place in `src`.
///
/// Returns error if there's a lone `\r` in the string
fn normalize_newlines(src: &mut String) {
    if !src.as_bytes().contains(&b'\r') {
        return;
    }

    // We replace `\r\n` with `\n` in-place, which doesn't break utf-8 encoding.
    // While we *can* call `as_mut_vec` and do surgery on the live string
    // directly, let's rather steal the contents of `src`. This makes the code
    // safe even if a panic occurs.

    let mut buf = std::mem::take(src).into_bytes();
    let mut gap_len = 0;
    let mut tail = buf.as_mut_slice();
    loop {
        let idx = match find_crlf(&tail[gap_len..]) {
            None => tail.len(),
            Some(idx) => idx + gap_len,
        };
        tail.copy_within(gap_len..idx, 0);
        tail = &mut tail[idx - gap_len..];
        if tail.len() == gap_len {
            break;
        }
        gap_len += 1;
    }

    // Account for removed `\r`.
    // After `set_len`, `buf` is guaranteed to contain utf-8 again.
    let new_len = buf.len() - gap_len;
    unsafe {
        buf.set_len(new_len);
        *src = String::from_utf8_unchecked(buf);
    }

    fn find_crlf(src: &[u8]) -> Option<usize> {
        let mut search_idx = 0;
        while let Some(idx) = find_cr(&src[search_idx..]) {
            if src[search_idx..].get(idx + 1) != Some(&b'\n') {
                search_idx += idx + 1;
                continue;
            }
            return Some(search_idx + idx);
        }
        None
    }

    fn find_cr(src: &[u8]) -> Option<usize> {
        src.iter().position(|&b| b == b'\r')
    }
}

fn snapshot_type_errors(path: &Path, source: &str) -> String {
    let result = run_frontend(source);
    if result
        .errors()
        .any(|e| matches!(e, CheckError::ParseError(_)))
    {
        panic!(
            "{} was expected to fail with a type error, but had a parse error instead",
            path.display()
        )
    }
    if !result.has_errors() {
        panic!("{} was expected to fail, but didn't", path.display())
    }

    let mut err_buf = String::new();
    for error in result.errors() {
        write!(
            &mut err_buf,
            "{}",
            error.display(source, &result.ctx, false)
        )
        .unwrap();
    }
    err_buf
}

#[test]
fn test_type_errors() {
    glob!("type_errors", "*.nemo", |path| {
        let input = {
            let mut input = fs::read_to_string(path).unwrap();
            normalize_newlines(&mut input);
            input
        };
        let output = {
            let mut output = snapshot_type_errors(path, &input);
            normalize_newlines(&mut output);
            output
        };
        assert_snapshot!(output)
    });
}

fn snapshot_parse_tree(input: &str) -> String {
    let parse = parse_prog(input);
    let tree = parse.debug_tree();
    let (_, errors) = parse.take();
    let errors = errors
        .into_iter()
        .map(|e| format!("{}", e.display(input, false)))
        .collect::<Vec<_>>()
        .join("\n");
    format!("{}\n\n{}", tree, errors)
}

#[test]
fn test_parsing() {
    glob!("parsing/*.nemo", |path| {
        let input = {
            let mut input = fs::read_to_string(path).unwrap();
            normalize_newlines(&mut input);
            input
        };
        let output = {
            let mut output = snapshot_parse_tree(&input);
            normalize_newlines(&mut output);
            output
        };
        assert_snapshot!(output);
    });
}
