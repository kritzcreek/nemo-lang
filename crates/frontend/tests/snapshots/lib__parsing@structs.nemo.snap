---
source: crates/frontend/tests/lib.rs
assertion_line: 129
expression: output
input_file: crates/frontend/tests/parsing/structs.nemo
---
Root@0..165
  Module@0..165
    TopGlobal@0..18
      GLOBAL_KW@0..6 "global"
      SPACE@6..7 " "
      IDENT@7..8 "_"
      SPACE@8..9 " "
      EQUALS@9..10 "="
      SPACE@10..11 " "
      EStruct@11..18
        UPPER_IDENT@11..14 "Vec"
        SPACE@14..15 " "
        L_BRACE@15..16 "{"
        SPACE@16..17 " "
        R_BRACE@17..18 "}"
    TopGlobal@18..44
      LINEFEED@18..19 "\n"
      GLOBAL_KW@19..25 "global"
      SPACE@25..26 " "
      IDENT@26..27 "_"
      SPACE@27..28 " "
      EQUALS@28..29 "="
      SPACE@29..30 " "
      EStruct@30..44
        UPPER_IDENT@30..33 "Vec"
        SPACE@33..34 " "
        L_BRACE@34..35 "{"
        SPACE@35..36 " "
        EStructField@36..43
          IDENT@36..37 "x"
          SPACE@37..38 " "
          EQUALS@38..39 "="
          SPACE@39..40 " "
          ELit@40..43
            LitInt@40..43
              INT_LIT@40..42 "10"
              SPACE@42..43 " "
        R_BRACE@43..44 "}"
    TopGlobal@44..79
      LINEFEED@44..45 "\n"
      GLOBAL_KW@45..51 "global"
      SPACE@51..52 " "
      IDENT@52..53 "_"
      SPACE@53..54 " "
      EQUALS@54..55 "="
      SPACE@55..56 " "
      EStruct@56..79
        UPPER_IDENT@56..59 "Vec"
        SPACE@59..60 " "
        L_BRACE@60..61 "{"
        SPACE@61..62 " "
        EStructField@62..68
          IDENT@62..63 "x"
          SPACE@63..64 " "
          EQUALS@64..65 "="
          SPACE@65..66 " "
          ELit@66..68
            LitInt@66..68
              INT_LIT@66..68 "10"
        COMMA@68..69 ","
        SPACE@69..70 " "
        EStructField@70..78
          IDENT@70..71 "y"
          SPACE@71..72 " "
          EQUALS@72..73 "="
          SPACE@73..74 " "
          ELit@74..78
            LitFloat@74..78
              FLOAT_LIT@74..77 "1.0"
              SPACE@77..78 " "
        R_BRACE@78..79 "}"
    TopGlobal@79..115
      LINEFEED@79..80 "\n"
      GLOBAL_KW@80..86 "global"
      SPACE@86..87 " "
      IDENT@87..88 "_"
      SPACE@88..89 " "
      EQUALS@89..90 "="
      SPACE@90..91 " "
      EStruct@91..115
        UPPER_IDENT@91..94 "Vec"
        SPACE@94..95 " "
        L_BRACE@95..96 "{"
        SPACE@96..97 " "
        EStructField@97..103
          IDENT@97..98 "x"
          SPACE@98..99 " "
          EQUALS@99..100 "="
          SPACE@100..101 " "
          ELit@101..103
            LitInt@101..103
              INT_LIT@101..103 "10"
        COMMA@103..104 ","
        SPACE@104..105 " "
        EStructField@105..112
          IDENT@105..106 "y"
          SPACE@106..107 " "
          EQUALS@107..108 "="
          SPACE@108..109 " "
          ELit@109..112
            LitFloat@109..112
              FLOAT_LIT@109..112 "1.0"
        COMMA@112..113 ","
        SPACE@113..114 " "
        R_BRACE@114..115 "}"
    TopGlobal@115..130
      LINEFEED@115..116 "\n"
      GLOBAL_KW@116..122 "global"
      SPACE@122..123 " "
      IDENT@123..124 "_"
      SPACE@124..125 " "
      EQUALS@125..126 "="
      SPACE@126..127 " "
      EStructIdx@127..130
        EVar@127..128
          IDENT@127..128 "v"
        DOT@128..129 "."
        IDENT@129..130 "x"
    TopGlobal@130..147
      LINEFEED@130..131 "\n"
      GLOBAL_KW@131..137 "global"
      SPACE@137..138 " "
      IDENT@138..139 "_"
      SPACE@139..140 " "
      EQUALS@140..141 "="
      SPACE@141..142 " "
      EStructIdx@142..147
        EStructIdx@142..145
          EVar@142..143
            IDENT@142..143 "v"
          DOT@143..144 "."
          IDENT@144..145 "x"
        DOT@145..146 "."
        IDENT@146..147 "y"
    TopGlobal@147..165
      LINEFEED@147..148 "\n"
      GLOBAL_KW@148..154 "global"
      SPACE@154..155 " "
      IDENT@155..156 "_"
      SPACE@156..157 " "
      EQUALS@157..158 "="
      SPACE@158..159 " "
      EStructIdx@159..165
        EArrayIdx@159..163
          EVar@159..160
            IDENT@159..160 "v"
          L_BRACKET@160..161 "["
          ELit@161..162
            LitInt@161..162
              INT_LIT@161..162 "0"
          R_BRACKET@162..163 "]"
        DOT@163..164 "."
        IDENT@164..165 "y"
